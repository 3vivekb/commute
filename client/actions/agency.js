import fetchAction from '@conveyal/woonerf/fetch'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addLocally = createAction('add agency')
export const createAgency = (newAgency) => {
  newAgency.id = uuid.v4()
  newAgency.organizations = []
  return [
    addLocally(newAgency),
    push(`/agency/${newAgency.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete agency')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/agency/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteAgency = (id) => [deleteLocally(id), push('/')]

export const loadAgencies = () => fetchAction({
  url: '/api/agency',
  next: (res, err) => {
    if (!err) {
      return setAllLocally(res)
    }
  }
})

export const updateAgency = (agency) => [
  setLocally(agency),
  push(`/agency/${agency.id}`)
]
/* const updateOnServer = (agency) =>
  serverAction({
    url: `/api/agency/${agency.id}`,
    options: {
      body: JSON.stringify(agency),
      method: 'put'
    }
  }) */  // TODO: update on server

export const setLocally = createAction('set agency')
export const setAllLocally = createAction('set agencies')
