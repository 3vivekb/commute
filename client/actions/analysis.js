import moment from 'moment'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addAnalysis = createAction('add analysis')
export const createAnalysis = (newAnalysis, organizationId) => {
  newAnalysis.id = uuid.v4()
  newAnalysis.lastRunDateTime = moment().unix()
  newAnalysis.trips = []
  return [
    addAnalysis({ analysis: newAnalysis, organizationId }),
    push(`/organizations/${organizationId}/analysis/${newAnalysis.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete analysis')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/analysis/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteAnalysis = (id, organzationId) => [
  deleteLocally({ analysisId: id, organzationId }),
  push(`/organizations/${organzationId}`)
]
