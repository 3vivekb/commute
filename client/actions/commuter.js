import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addCommuter = createAction('add commuter')
export const createCommuter = (commuter, groupId) => {
  commuter.groupId = groupId
  commuter.id = uuid.v4()
  return [
    addCommuter(commuter),
    push(`/group/${groupId}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete commuter')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/commuter/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteCommuter = (commuterId, groupId) => [
  deleteLocally(commuterId),
  push(`/group/${groupId}`)
]

const updateLocally = createAction('update commuter')
/* const updateOnServer = (id) =>
  serverAction({
    url: `/api/commuter/${id}`,
    options: {
      method: 'update'
    }
  }) */ // TODO update on server
export const updateCommuter = (commuter, groupId) => [
  updateLocally(commuter),
  push(`/group/${groupId}`)
]
