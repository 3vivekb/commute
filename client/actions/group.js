import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

// site stuff
const addGroup = createAction('add group')
export const createGroup = (newGroup, organizationId) => {
  newGroup.id = uuid.v4()
  return [
    addGroup({ group: newGroup, organizationId }),
    push(`/organizations/${organizationId}/group/${newGroup.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete group')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/group/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteGroup = (id, organizationId) => [
  deleteLocally(id),
  push(`/organizations/${organizationId}`)
]
