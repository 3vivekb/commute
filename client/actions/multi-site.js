import update from 'react-addons-update'

import makeGenericModelActions from '../utils/actions'

const actions = makeGenericModelActions({
  commands: {
    'Collection GET': {},
    'Collection POST': {},
    'DELETE': {
      redirectionStrategy: 'toHome'
    },
    'GET': {},
    'PUT': {}
  },
  pluralName: 'multi-sites',
  singularName: 'multi-site'
})

actions.deleteSiteFromMultiSites = function ({multiSites, siteId}) {
  const actionsToDispatch = []
  multiSites.forEach((multiSite) => {
    const siteIdx = multiSite.sites.indexOf(siteId)
    if (siteIdx > -1) {
      // remove site from array
      const updatedMultiSite = update(multiSite, {
        sites: {
          $splice: [[siteIdx, 1]]
        }
      })

      // dispatch update action for multiSite
      actionsToDispatch.push(actions.update(updatedMultiSite, 'no-redirect'))
    }
  })

  return actionsToDispatch
}

export default actions
