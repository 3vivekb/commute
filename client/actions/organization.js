import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addOrganization = createAction('add organization')
export const createOrganization = (newOrganization) => {
  newOrganization.id = uuid.v4()
  newOrganization.sites = []
  newOrganization.groups = []
  return [
    addOrganization(newOrganization),
    push('/')
  ] // TODO save to server
}

// site stuff
const addSite = createAction('add site')
export const createSite = (newSite) => {
  newSite.id = uuid.v4()
  return [
    addSite(newSite),
    push(`/organizations/${newSite.organizationId}`)
  ] // TODO save to server
}
