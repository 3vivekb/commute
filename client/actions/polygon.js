import makeGenericModelActions from '../utils/actions'

const actions = makeGenericModelActions({
  commands: {
    'Collection DELETE': {},
    'Collection GET': {}
  },
  pluralName: 'polygons',
  singularName: 'polygon'
})

export default actions
