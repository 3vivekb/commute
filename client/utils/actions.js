import fetchAction from '@conveyal/woonerf/fetch'
import qs from 'qs'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'

/**
 * Make generic model actions that communicate with a rest server
 *
 * @param  {Object} cfg An object with the following parameters:
 * - commands: An object with the keys being commands and the corresponding cfg of each command
 * -- redirectionStrategy: The redirection strategy to use upon success.
 *  Can be one of following: 'toEntity', 'toParent', 'toHome', undefined (no redirection)
 *  Defaults are as follows:
 *    Collection GET, GET: null
 *    Collection POST, PUT: toEntity
 *    DELETE: toParent
 * - parentKey: Key in model of parent's id
 *   (must be provided if using toParent redirection strategy)
 * - parentName: Name of app route to parent view
 *   (must be provided if using toParent redirection strategy)
 * - pluralName: Plural name of the model
 * - singularName: Singular name of the model
 * @return {Object}     The actions
 */
const makeGenericModelActions = (cfg) => {
  const {commands, parentKey, parentName, pluralName, singularName} = cfg
  const actions = {}

  // make local set actions
  const addLocally = createAction(`add ${singularName}`)
  const deleteLocally = createAction(`delete ${singularName}`)
  const setLocally = createAction(`set ${singularName}`)
  const setAllLocally = createAction(`set ${pluralName}`)

  const baseEndpoint = `/api/${singularName}`

  const redirectionStrategies = {
    'toEntity': (entity) => {
      return push(`/${singularName}/${entity._id}`)
    },
    'toHome': () => {
      return push('/')
    },
    'toParent': (entity) => {
      return push(`/${parentName}/${entity[parentKey]}`)
    }
  }

  /**
   * Do a redirect if needed
   *
   * @param  {Object} cfg Paremters as follows:
   *   - actions: The actions to append to
   *   - endpointCfg: The endpoint config
   *   - defaultStrategy: The default redire strategy to use (default: undefined)
   *   - redirectArgs: Arguments to provide to the redirectionStrategy
   */
  const doRedirectIfNecessary = (cfg) => {
    const redirectionStrategy = (cfg.endpointCfg.redirectionStrategy !== undefined
      ? cfg.endpointCfg.redirectionStrategy
      : cfg.defaultStrategy)

    if (redirectionStrategy) {
      const redirect = redirectionStrategies[redirectionStrategy](cfg.redirectArgs)
      if (redirect) {
        cfg.actions.push(redirect)
      }
    }
  }

  if (commands['Collection GET']) {
    actions.loadAll = (queryParams) => {
      // only include filteredKeys in querystring
      let queryString = ''
      if (queryParams) {
        queryString = qs.stringify(queryParams)
        if (queryString.length !== 0) {
          queryString = '?' + queryString
        }
      }

      // make request
      return fetchAction({
        next: (res, err) => {
          if (!err) {
            return setAllLocally(res.value)
          }
          // TODO handle error
        },
        url: baseEndpoint + queryString
      })
    }
  }

  if (commands['Collection POST']) {
    const endpointCfg = commands['Collection POST']
    actions.create = (newEntity) => {
      return fetchAction({
        next: (res, err) => {
          if (err) {
            // TODO handle error
          } else {
            const createdEntity = res.value
            const actions = [
              addLocally(createdEntity)
            ]
            doRedirectIfNecessary({
              actions,
              endpointCfg,
              defaultStrategy: 'toEntity',
              redirectArgs: createdEntity
            })
            return actions
          }
        },
        options: {
          body: newEntity,
          method: 'POST'
        },
        url: baseEndpoint
      })
    }
  }

  if (commands['DELETE']) {
    const endpointCfg = commands['DELETE']
    actions.delete = (entity) => {
      return fetchAction({
        next: (res, err) => {
          if (err) {
            // TODO handle error
          } else {
            const actions = [
              deleteLocally(entity._id)
            ]
            doRedirectIfNecessary({
              actions,
              endpointCfg,
              defaultStrategy: 'toParent',
              redirectArgs: entity
            })
            return actions
          }
        },
        options: {
          method: 'DELETE'
        },
        url: `${baseEndpoint}/${entity._id}`
      })
    }
  }

  if (commands['GET']) {
    actions.loadOne = (id) => fetchAction({
      next: (res, err) => {
        if (!err) {
          return setLocally(res.value)
        }
        // TODO handle error
      },
      url: `${baseEndpoint}/${id}`
    })
  }

  if (commands['PUT']) {
    const endpointCfg = commands['PUT']
    actions.update = (entity) => fetchAction({
      next: (res, err) => {
        if (err) {
          // TODO handle error
        } else {
          const updatedEntity = res.value
          const actions = [
            setLocally(updatedEntity)
          ]
          doRedirectIfNecessary({
            actions,
            endpointCfg,
            defaultStrategy: 'toEntity',
            redirectArgs: updatedEntity
          })
          return actions
        }
      },
      options: {
        body: entity,
        method: 'PUT'
      },
      url: `${baseEndpoint}/${entity._id}`
    })
  }

  return actions
}

export default makeGenericModelActions
