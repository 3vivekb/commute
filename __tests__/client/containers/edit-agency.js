/* global describe, expect, it */

import {mount} from 'enzyme'
import {mountToJson} from 'enzyme-to-json'
import React from 'react'
import {Provider} from 'react-redux'

import {expectCreateAction, expectDeleteAgency} from '../../test-utils/actions'
import {makeMockStore, mockStores} from '../../test-utils/mock-data.js'

import EditAgency from '../../../client/containers/edit-agency'

describe('Container > EditAgency', () => {
  it('Create/Edit Agency View loads (create mode)', () => {
    const mockStore = makeMockStore(mockStores.init)

    // mount component
    const tree = mount(
      <Provider store={mockStore}>
        <EditAgency
          params={{}}
          />
      </Provider>
    )
    expect(mountToJson(tree.find(EditAgency))).toMatchSnapshot()
  })

  it('Create/Edit Agency View loads in edit mode', () => {
    const mockStore = makeMockStore(mockStores.withBlankAgency)

    // mount component
    const tree = mount(
      <Provider store={mockStore}>
        <EditAgency
          params={{agencyId: 'agency-1'}}
          />
      </Provider>
    )
    expect(mountToJson(tree.find(EditAgency))).toMatchSnapshot()
  })

  it('Create agency', () => {
    const mockStore = makeMockStore(mockStores.init)

    // mount component
    const tree = mount(
      <Provider store={mockStore}>
        <EditAgency
          params={{}}
          />
      </Provider>
    )

    // give each text field some input
    tree.find('input').map((input) => input.simulate('change', {target: {value: 'My new value'}}))

    // submit form
    tree.find('form').find('button').simulate('click')

    // expect create action
    expectCreateAction(mockStore.getActions())
  })

  it('Update agency', () => {
    const mockStore = makeMockStore(mockStores.withBlankAgency)

    // mount component
    const tree = mount(
      <Provider store={mockStore}>
        <EditAgency
          params={{agencyId: 'agency-1'}}
          />
      </Provider>
    )

    // give each text field some input
    tree.find('input').map((input) => input.simulate('change', {target: {value: 'My new value'}}))

    // submit form
    tree.find('form').find('button').first().simulate('click')

    // expect update action
    const actions = mockStore.getActions()
    expect(actions.length).toBe(2)
    expect(actions).toMatchSnapshot()
  })

  it('Delete agency', () => {
    const mockStore = makeMockStore(mockStores.withBlankAgency)
    window.confirm = () => true

    // Given a logged-in user is viewing the Create/Edit Agency View
    // mount component
    const tree = mount(
      <Provider store={mockStore}>
        <EditAgency
          params={{agencyId: 'agency-1'}}
          />
      </Provider>
    )

    // When the user clicks the delete button for an existing agency
    // And the user confirms the Confirm Deletion dialog
    const deleteButton = tree.find('button').last()
    deleteButton.simulate('click')

    expectDeleteAgency(mockStore.getActions())
  })
})
