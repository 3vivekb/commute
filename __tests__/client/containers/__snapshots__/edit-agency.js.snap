exports[`Container > EditAgency Create agency 1`] = `
Object {
  "payload": Object {
    "name": "My new value",
    "organizations": Array [],
  },
  "type": "add agency",
}
`;

exports[`Container > EditAgency Create agency 2`] = `
Object {
  "payload": Object {
    "args": Array [
      "/agency/new-entity-id",
    ],
    "method": "push",
  },
  "type": "@@router/CALL_HISTORY_METHOD",
}
`;

exports[`Container > EditAgency Create/Edit Agency View loads (create mode) 1`] = `
<Connect(EditAgency)
  params={Object {}}>
  <EditAgency
    create={[Function]}
    delete={[Function]}
    editMode={false}
    params={Object {}}
    update={[Function]}>
    <Grid
      bsClass="container"
      componentClass="div"
      fluid={false}>
      <div
        className="container">
        <Row
          bsClass="row"
          componentClass="div">
          <div
            className="row">
            <Col
              bsClass="col"
              componentClass="div"
              xs={12}>
              <div
                className="col-xs-12">
                <h3>
                  <span>
                    Create Agency
                  </span>
                  <Button
                    active={false}
                    block={false}
                    bsClass="btn"
                    bsStyle="default"
                    className="pull-right"
                    disabled={false}>
                    <button
                      className="pull-right btn btn-default"
                      disabled={false}
                      type="button">
                      <Link
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to="/">
                        <a
                          onClick={[Function]}
                          style={Object {}}>
                          <Icon
                            type="arrow-left">
                            <i
                              className="fa fa-arrow-left fa-fw " />
                          </Icon>
                          <span>
                            Back
                          </span>
                        </a>
                      </Link>
                    </button>
                  </Button>
                </h3>
                <form>
                  <FieldGroup
                    label="Name"
                    name="name"
                    onChange={[Function]}
                    placeholder="Enter name"
                    type="text"
                    value={undefined}>
                    <FormGroup
                      bsClass="form-group"
                      controlId="group-item-name">
                      <div
                        className="form-group">
                        <ControlLabel
                          bsClass="control-label"
                          srOnly={false}>
                          <label
                            className="control-label"
                            htmlFor="group-item-name">
                            Name
                          </label>
                        </ControlLabel>
                        <FormControl
                          bsClass="form-control"
                          componentClass="input"
                          onChange={[Function]}
                          placeholder="Enter name"
                          type="text"
                          value="">
                          <input
                            className="form-control"
                            id="group-item-name"
                            onChange={[Function]}
                            placeholder="Enter name"
                            type="text"
                            value="" />
                        </FormControl>
                      </div>
                    </FormGroup>
                  </FieldGroup>
                  <Button
                    active={false}
                    block={false}
                    bsClass="btn"
                    bsStyle="success"
                    disabled={false}
                    onClick={[Function]}>
                    <button
                      className="btn btn-success"
                      disabled={false}
                      onClick={[Function]}
                      type="button">
                      Create
                    </button>
                  </Button>
                </form>
              </div>
            </Col>
          </div>
        </Row>
      </div>
    </Grid>
  </EditAgency>
</Connect(EditAgency)>
`;

exports[`Container > EditAgency Create/Edit Agency View loads in edit mode 1`] = `
<Connect(EditAgency)
  params={
    Object {
      "agencyId": "agency-1",
    }
  }>
  <EditAgency
    agency={
      Object {
        "id": "agency-1",
        "name": "Mock Agency",
        "organizations": Array [],
      }
    }
    create={[Function]}
    delete={[Function]}
    editMode={true}
    params={
      Object {
        "agencyId": "agency-1",
      }
    }
    update={[Function]}>
    <Grid
      bsClass="container"
      componentClass="div"
      fluid={false}>
      <div
        className="container">
        <Row
          bsClass="row"
          componentClass="div">
          <div
            className="row">
            <Col
              bsClass="col"
              componentClass="div"
              xs={12}>
              <div
                className="col-xs-12">
                <h3>
                  <span>
                    Edit Agency
                  </span>
                  <Button
                    active={false}
                    block={false}
                    bsClass="btn"
                    bsStyle="default"
                    className="pull-right"
                    disabled={false}>
                    <button
                      className="pull-right btn btn-default"
                      disabled={false}
                      type="button">
                      <Link
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to="/">
                        <a
                          onClick={[Function]}
                          style={Object {}}>
                          <Icon
                            type="arrow-left">
                            <i
                              className="fa fa-arrow-left fa-fw " />
                          </Icon>
                          <span>
                            Back
                          </span>
                        </a>
                      </Link>
                    </button>
                  </Button>
                </h3>
                <form>
                  <FieldGroup
                    label="Name"
                    name="name"
                    onChange={[Function]}
                    placeholder="Enter name"
                    type="text"
                    value="Mock Agency">
                    <FormGroup
                      bsClass="form-group"
                      controlId="group-item-name">
                      <div
                        className="form-group">
                        <ControlLabel
                          bsClass="control-label"
                          srOnly={false}>
                          <label
                            className="control-label"
                            htmlFor="group-item-name">
                            Name
                          </label>
                        </ControlLabel>
                        <FormControl
                          bsClass="form-control"
                          componentClass="input"
                          onChange={[Function]}
                          placeholder="Enter name"
                          type="text"
                          value="Mock Agency">
                          <input
                            className="form-control"
                            id="group-item-name"
                            onChange={[Function]}
                            placeholder="Enter name"
                            type="text"
                            value="Mock Agency" />
                        </FormControl>
                      </div>
                    </FormGroup>
                  </FieldGroup>
                  <Button
                    active={false}
                    block={false}
                    bsClass="btn"
                    bsStyle="warning"
                    disabled={false}
                    onClick={[Function]}>
                    <button
                      className="btn btn-warning"
                      disabled={false}
                      onClick={[Function]}
                      type="button">
                      Update
                    </button>
                  </Button>
                  <Button
                    active={false}
                    block={false}
                    bsClass="btn"
                    bsStyle="danger"
                    disabled={false}
                    onClick={[Function]}>
                    <button
                      className="btn btn-danger"
                      disabled={false}
                      onClick={[Function]}
                      type="button">
                      Delete
                    </button>
                  </Button>
                </form>
              </div>
            </Col>
          </div>
        </Row>
      </div>
    </Grid>
  </EditAgency>
</Connect(EditAgency)>
`;

exports[`Container > EditAgency Delete agency 1`] = `
Array [
  Object {
    "payload": "agency-1",
    "type": "delete agency",
  },
  Object {
    "payload": Object {
      "args": Array [
        "/",
      ],
      "method": "push",
    },
    "type": "@@router/CALL_HISTORY_METHOD",
  },
]
`;

exports[`Container > EditAgency Update agency 1`] = `
Array [
  Object {
    "payload": Object {
      "id": "agency-1",
      "name": "My new value",
      "organizations": Array [],
    },
    "type": "set agency",
  },
  Object {
    "payload": Object {
      "args": Array [
        "/agency/agency-1",
      ],
      "method": "push",
    },
    "type": "@@router/CALL_HISTORY_METHOD",
  },
]
`;
